Windows Native `taste-tester upload`

svmastersamurai opened the pull request:
These changes will let you `upload` a chef development repo to `taste-tester` natively on Windows.

* Windows cannot run `chef-zero` with the `--daemon` flag, so instead we create a scheduled task that will spawn a `chef-zero` process. The side-effect of this is that a window will remain open for the duration of the testing.
* Added a Windows module to contain windows specific code
* Shifted order of taste-tester issuing the `@state.wipe` when in `stop` mode.
* Added `pid` method to `State` class so it can infer the chef-zero process id
* Added `windows?` method to `Server` class so it only executes windows specific code.

jaymzh commented on lib/taste_tester/server.rb:
```diff
@@ -84,14 +85,14 @@ def start
       end
       @state.wipe
-      write_config
       start_chef_zero
```
Out of curiosity why ?

jaymzh replied:
still curious why you switched the order here.

jaymzh commented on lib/taste_tester/server.rb:
```diff
@@ -152,19 +153,30 @@ def start_chef_zero
+      if windows?
+        extend ::TasteTester::Windows
+        stop_win_chef_zero_server
+        @state.wipe
```
you do this in `stop` above, no need to do it here...

svmastersamurai replied:
👍

jaymzh commented on lib/taste_tester/windows.rb:
```diff
@@ -0,0 +1,64 @@
+module TasteTester
+  module Windows
+    PATH_TO_WRAPPER_SCRIPT =
+      File.absolute_path(File.join(
```
OK, so if I understand this all correctly, the only reason for the wrapper script is so that when the scheduled task gets run, we shove the PID into the statefile?

That's... terrible. I mean it's a neat hack... but Jesus, Microsoft.

svmastersamurai replied:
Yeah the main problem is that Windows can't use the `--daemon` flag. There may be a better way to do this, but yeah I agree it is a big ball of poo 😢

jaymzh commented on scripts/win_chef_zero_server.ps1:
```diff
@@ -0,0 +1,27 @@
+# so it can be referenced in the plugin.
+$state = Get-Content ~/.chef/taste-tester-ref.json | ConvertFrom-JSON
+$state | Add-Member -Type NoteProperty -Name pid -Value $proc.id -Force
+$state |
```
There's a race-condition here... to be clear. but it's probably ... OK. We should at least add a comment to that effect though.

svmastersamurai replied:
Another way is to make the PID placement for all platforms. I was purposefully trying to keep the changes restricted to the bare minimum to get Windows operational but I can revise this to make it less hacky if need be!

svmastersamurai commented:
Alrighty, moved the `@state` mutilation into Ruby with the help of the `wmi-lite` gem. And uh... dealt with merging latest cut of `master`. I don't think I cherry-picked things right, but the feedback is making my additions look much cleaner! 🍨

jaymzh commented:
Oh god you merged somethign somewhere. Please rebase so the diffs are ontop of master :)

jaymzh commented on lib/taste_tester/windows.rb:
```diff
@@ -0,0 +1,90 @@
+      ps_opts = '-noprofile -noninteractive'
+      script_opts = '-UseSSL' if @state.ssl
+      cmd = format(
+        'powershell.exe %s -file %s -CZHost %s -CZPort %s %s',
```
now that we don't have the crazy pid thing, can't we just run chef-zero directly here? Do we need the wrapper script?

jaymzh commented:
Oh you'll probably have to force-push to your branch...

svmastersamurai commented:
Ah, i'll give it a try! Should get around to updates tomorrowish

svmastersamurai commented:
@jaymzh tbh I couldn't tell you _exactly why_ I changed the ordering aside from it was necessary at the time to get the first version working. I managed to refactor the code a bit and all those calls are now back to their original form.

This iteration removed all the powershell stuff, it looks like `Mixlib::ShellOut` was the culprit: it would just wait for the process to finish once it was created. If you're cool with using `Process.spawn` the end result is all the same, and with much less clowniness 😎 

Cleaned out my `git` log too, we should be 👍  there

jaymzh commented:
LGTM. I'll merge when tests pass.

jaymzh commented:
Thanks!
